src/solvers/dg_unstructured_quad/containers_2d.jl

```diff
@@ -0,0 +1,320 @@
+
+# Container data structure (structure-of-arrays style) for DG elements on curved unstructured mesh
+struct UnstructuredElementContainer2D{RealT<:Real, uEltype<:Real, NVARS, POLYDEG}
+  node_coordinates   ::Array{RealT, 4} # [ndims, nnodes, nnodes, nelement]
+  X_xi               ::Array{RealT, 3} # [nnodes, nnodes, nelement]
+  X_eta              ::Array{RealT, 3} # [nnodes, nnodes, nelement]
+  Y_xi               ::Array{RealT, 3} # [nnodes, nnodes, nelement]
+  Y_eta              ::Array{RealT, 3} # [nnodes, nnodes, nelement]
+  inverse_jacobian   ::Array{RealT, 3} # [nnodes, nnodes, nelement]
+  normals            ::Array{RealT, 4} # [ndims, nnodes, local sides, nelement]
+  tangents           ::Array{RealT, 4} # [ndims, nnodes, local sides, nelement]
+  scaling            ::Array{RealT, 3} # [nnodes, local sides, nelement]
+  surface_flux_values::Array{RealT, 4} # [variables, nnodes, local sides, elements]
+end
+
+
+# construct an empty curved element container to be filled later with geometries in the
+# unstructured mesh constructor
+function UnstructuredElementContainer2D{RealT, uEltype, NVARS, POLYDEG}(capacity::Int64) where {RealT<:Real, uEltype<:Real, NVARS, POLYDEG}
+
+  nnodes = POLYDEG + 1
+  nan = convert(RealT, NaN)
+
+  node_coordinates    = fill(nan, (2, nnodes, nnodes, capacity))
+  X_xi                = fill(nan, (nnodes, nnodes, capacity))
+  X_eta               = fill(nan, (nnodes, nnodes, capacity))
+  Y_xi                = fill(nan, (nnodes, nnodes, capacity))
+  Y_eta               = fill(nan, (nnodes, nnodes, capacity))
+  inverse_jacobian    = fill(nan, (nnodes, nnodes, capacity))
+  normals             = fill(nan, (2, nnodes, 4, capacity))
+  tangents            = fill(nan, (2, nnodes, 4, capacity))
+  scaling             = fill(nan, (nnodes, 4, capacity))
+  surface_flux_values = fill(nan, (NVARS, nnodes, 4, capacity))
+
+  return UnstructuredElementContainer2D{RealT, uEltype, NVARS, POLYDEG}(node_coordinates,
+                                                                        X_xi, X_eta, Y_xi, Y_eta,
+                                                                        inverse_jacobian,
+                                                                        normals, tangents, scaling,
+                                                                        surface_flux_values)
+end
+
+
+@inline nelements(elements::UnstructuredElementContainer2D) = size(elements.inverse_jacobian, 3)
+
+@inline eachelement(elements::UnstructuredElementContainer2D) = Base.OneTo(nelements(elements))
+
+# TODO: I do not fully understand what this is doing. What exactly is uEltype and why do we use it?
+Base.eltype(::UnstructuredElementContainer2D{RealT, uEltype}) where {RealT, uEltype} = uEltype
+
+
+function init_elements(RealT, mesh, dg_nodes, nvars, poly_deg)
+
+  elements = UnstructuredElementContainer2D{RealT, RealT, nvars, poly_deg}(mesh.n_elements)
```

Comment:
When the type instability of `GammaCurves` is fixed, you might want to introduce a function barrier `init_elements!(elements, ...)` here similarly to other setups to improve performance.